pkg/server/backing_image.go

```diff
@@ -75,49 +75,47 @@ func NewBackingImage(name, url, uuid, diskPathOnHost, diskPathInContainer string
 		),
 		lock:       &sync.RWMutex{},
 		downloader: downloader,
+		updateCh:   updateCh,
 	}
 }
 
 func GetBackingImageDirectoryName(biName, biUUID string) string {
 	return fmt.Sprintf("%s-%s", biName, biUUID)
 }
 
-func (bi *BackingImage) SetUpdateChannel(updateCh chan interface{}) {
-	bi.updateCh = updateCh
-}
-
-func IntroduceBackingImage(name, url, uuid, diskPathOnHost, diskPathInContainer, state string, size int64, downloader Downloader) *BackingImage {
-	bi := NewBackingImage(name, url, uuid, diskPathOnHost, diskPathInContainer, downloader)
+func (bi *BackingImage) Pull() (resp *rpc.BackingImageResponse, err error) {
 	bi.lock.Lock()
-	defer bi.lock.Unlock()
-	if name == "" || uuid == "" || diskPathOnHost == "" || size <= 0 || state != types.DownloadStateDownloaded {
-		bi.state = types.DownloadStateFailed
-	} else {
-		if err := bi.checkAndReuseBackingImageFileWithoutLock(); err != nil || bi.size != size {
-			bi.state = types.DownloadStateFailed
-		}
+	log := bi.log
+	log.Info("Backing Image: start to pull backing image")
+
+	if bi.state != "" {
+		state := bi.state
+		bi.lock.Unlock()
+		return nil, fmt.Errorf("invalid state %v for pulling", state)
 	}
-	return bi
-}
+	bi.state = types.DownloadStatePending
```

Comment:
Suggest to move the state changer after checking `bin.downloader.InitDownloading()` successfully, because we should not do any state if the caller can't take the ownership.